File: exercises/week5/series/sources-solution/print_series.py
from __future__ import print_function
import math
from dumper_series import DumperSeries
################################################################


class PrintSeries(DumperSeries):

    def __init__(self, series, maxiter, freq):

        DumperSeries.__init__(self, series)

        self.maxiter = maxiter
        self.freq = freq

    def __del__(self):
        pass

    def dump(self):
        out = ""
        for i in range(1, int(self.maxiter/self.freq)):
            res = self.series.compute(i*self.freq-1)
            res2 = self.series.compute(i*self.freq)

            my_format = '{:.' + str(self.precision) + 'e}'
            out += str(i*self.freq) + " "
            out += my_format.format(res) + " "
            out += my_format.format(res2 - res)

            try:
                out += " " + my_format.format(
                    math.fabs(res2 - self.series.getAnalyticPrediction()))

            except Exception as e:
                pass

            out += "\n"
        return out

File: exercises/week6/series/sources-solution/src/print_series.cc
#include <cmath>
#include <iostream>
/* -------------------------------------------------------------------------- */
#include "print_series.hh"
/* -------------------------------------------------------------------------- */

PrintSeries::PrintSeries(Series &series, UInt freq)
    : DumperSeries(series), freq(freq) {}

/* -------------------------------------------------------------------------- */

void PrintSeries::dump(std::ostream &os) {
  for (UInt i = 1; i * this->freq < this->maxiter; ++i) {
    double res = series.compute(i * this->freq - 1);
    double res2 = series.compute(i * this->freq);

    os << std::setprecision(this->precision);
    os << std::scientific;

    os << i * this->freq << " " << res << " " << res2 - res;
    if (!std::isnan(series.getAnalyticPrediction())) {
      os << " " << std::abs(res2 - series.getAnalyticPrediction());
    }
    os << std::endl;
  }
}
